zcash_client_sqlite/src/wallet/sapling/commitment_tree.rs

```diff
@@ -214,3 +256,98 @@ pub(crate) fn checkpoint_count(conn: &rusqlite::Connection) -> Result<usize, Err
     })
     .map_err(Either::Right)
 }
+
+pub(crate) fn get_checkpoint<C: Deref<Target = Connection>>(
+    conn: &C,
+    checkpoint_id: BlockHeight,
+) -> Result<Option<Checkpoint>, Either<io::Error, rusqlite::Error>> {
+    let checkpoint_position = conn
+        .query_row(
+            "SELECT position
+                FROM sapling_tree_checkpoints
+                WHERE checkpoint_id = ?",
+            [u32::from(checkpoint_id)],
+            |row| {
+                row.get::<_, Option<u64>>(0)
+                    .map(|opt| opt.map(Position::from))
+            },
+        )
+        .optional()
+        .map_err(Either::Right)?;
+
+    let mut marks_removed = BTreeSet::new();
+    let mut stmt = conn
+        .prepare_cached(
+            "SELECT mark_removed_position
+                FROM sapling_tree_checkpoint_marks_removed
+                WHERE checkpoint_id = ?",
+        )
```

Comment:
Hm. When adding a checkpoint, the `marks_removed` set should always be empty (because you're never adding a checkpoint when you have information from the future).